### PR TITLE
BGP neighbor puppet support for XR

### PR DESCRIPTION
**Pull request by chburket**

### CODE REVIEW

**Comment by glennmatthews** [comment]
Please fix up your commit history (there are three commits by 'root' in your diffs).

File: README.md [hunk 1]
```diff
@@ -551,37 +551,37 @@ Description of the neighbor. Valid value is string.
...
 ##### `ebgp_multihop`
-Specify multihop TTL for a remote peer. Valid values are integers between 2  and 255, or keyword 'default' to disable this property.
+Specify multihop TTL for a remote peer. Valid values are integers between 2 and 255, or keyword 'default' to disable this property.
 
 ##### `local_as`
 Specify the local-as number for the eBGP neighbor. Valid values are String or Integer in ASPLAIN or ASDOT notation, or 'default', which means not to configure it.
 
 ##### `log_neighbor_changes`
-Specify wether or not to enable log messages for neighbor up/down event. Valid values are 'enable', to enable it, 'disable' to disable it, or 'inherit' to use the configuration in the cisco_bgp type.
+Specify wether or not to enable log messages for neighbor up/down event. Valid values are 'enable', to enable it, 'disable' to disable it, or 'inherit' to use the configuration in the cisco_bgp type. This property is not supported on IOS XR.
```
Comment (glennmatthews) [inline README.md]: While you're changing this, can you `s/wether/whether/`?

File: README.md [hunk 2]
```diff
@@ -551,37 +551,37 @@ Description of the neighbor. Valid value is string.
...
 ##### `maximum_peers`
-Specify Maximum number of peers for this neighbor prefix. Valid values are between 1 and 1000, or 'default', which does not impose the limit. This  attribute can only be configured if neighbor is in 'ip/prefix' format.
+Specify Maximum number of peers for this neighbor prefix. Valid values are between 1 and 1000, or 'default', which does not impose the limit. This attribute can only be configured if neighbor is in 'ip/prefix' format, and is therefore not supported on IOS XR.
 
 ##### `password`
 Specify the password for neighbor. Valid value is string.
 
 ##### `password_type`
-Specify the encryption type the password will use. Valid values are 'cleartext', '3des' or 'cisco_type_7' encryption, and 'default',which defaults to 'cleartext'.
+Specify the encryption type the password will use. Valid values are 'cleartext', '3des' or 'cisco_type_7' encryption, and 'default', which defaults to 'cleartext'.  Valid values for IOS XR are 'cleartext', 'md5', and 'default'.
```
Comment (glennmatthews) [inline README.md]: Probably need to insert 'for Nexus' into the second sentence of this line?

Is the default 'cleartext' for XR as well? It's not clear from the phrasing.

File: examples/demo_bgp.pp [hunk 3]
```diff
@@ -130,81 +130,148 @@
...
+      description            => 'my description',
+      connected_check        => true,
+      ebgp_multihop          => 2,
+      local_as               => 55.88,
+      remote_as              => 12,
+      shutdown               => true,
+      suppress_4_byte_as     => true,
+      timers_keepalive       => 90,
+      timers_holdtime        => 270,
+      update_source          => 'fastethernet1/1/1/1',
```
Comment (glennmatthews) [inline examples/demo_bgp.pp]: What platform are you testing on such that it has FastEthernet interfaces?

Comment (chburket) [inline examples/demo_bgp.pp]: The XR that is on the lab pod.  I just chose it from the options that were presented via 'update_source ?'.  Would you rather I change it to GigabitEthernet?

File: README.md (hunk 2 above)
Comment (chburket) [inline README.md]: Will fix.

File: README.md (hunk 1 above)
Comment (chburket) [inline README.md]: Yes

File: examples/demo_bgp.pp [hunk 4]
```diff
@@ -130,81 +130,148 @@
...
+    cisco_bgp_neighbor {'55.77 blue 3.3.3.3':
+      ensure      => present,
+
+      #Properties
+      description => 'config_for_rr_client',
+      remote_as   => 3604557,
+      shutdown    => false,
+    }
   }
+  elsif $operatingsystem == 'ios_xr' {
```
Comment (glennmatthews) [inline examples/demo_bgp.pp]: It would be worth looking into the various [conditional operators](http://docs.puppetlabs.com/puppet/4.3/reference/lang_conditional.html#selectors) to see if you can have OS-dependent logic and shared type declaration, something like:

```puppet
$capability_negotiation = $operatingsystem ? {
  'nexus' => true
  default => undef
}
$dynamic_capability = $operatingsystem ? {
  'nexus' => true
  default => undef
}

    # actual neighbor declaration is shared between both OS:
    cisco_bgp_neighbor {'55.77 blue 1.1.1.1':
      ensure                 => present,

      #Properties
      description            => 'my description',
      connected_check        => true,
      capability_negotiation => $capability_negotiation,
      dynamic_capability     => $dynamic_capability,
      ebgp_multihop          => 2,
      local_as               => 55.77,
...
```

**Comment by chburket** [comment]
Weird.  How do I change the author of a commit??


________________________________

File: tests/beaker_tests/bgpneighbor/test_bgpneighbor_password.rb [hunk 5]
```diff
@@ -119,9 +136,13 @@
     create_bgpneighbor_manifest(tests, id)
     test_manifest(tests, id)
 
-    tests[id][:desc] = '1.5 Verify password has been removed on the box'
-    test_show_cmd(tests, id, true)
+    tests[id][:desc] = '1.4 Verify password has been removed on the box'
+    tests[id][:resource] = { 'password' => UtilityLib::IGNORE_VALUE }
```
Comment (glennmatthews) [inline tests/beaker_tests/bgpneighbor/test_bgpneighbor_password.rb]: I like this. Nice approach!

**Comment by glennmatthews** [comment]
There's various ways to fixup the history. The easiest may be to do a 'git rebase -i' onto the parent branch and simply squash your four commits together into a single commit, then 'git push --force' to overwrite the record of your branch on github.

File: examples/demo_bgp.pp (hunk 3 above)
Comment (glennmatthews) [inline examples/demo_bgp.pp]: No, it's fine - I'd just expected that XR would reject it if the interface doesn't actually exist. If it's working with this, that's fine (and probably safer than pointing to any specific real interface).

**Comment by RichWellum** [comment]
@chburket 

$ git config --global user.name "John Doe"
$ git config --global user.email johndoe@example.com

File: examples/demo_bgp.pp (hunk 4 above)
Comment (chburket) [inline examples/demo_bgp.pp]: I used demo_interface.pp as an example.  Do you think it would be more clear to use selectors instead of conditional blocks?

Comment (glennmatthews) [inline examples/demo_bgp.pp]: As we discussed in scrum this morning - yes, I do. Will hopefully showcase the commonalities between both OS and highlight the differences more clearly.

File: README.md [hunk 6]
```diff
@@ -1146,7 +1146,7 @@ Manages a Cisco VLAN.
 ID of the Virtual LAN. Valid value is an integer.
 
 ##### `ensure`
-Determined wether the config should be present or not. Valid values are
+Determined whether the config should be present or not. Valid values are
```
Comment (glennmatthews) [inline README.md]: s/Determined/Determine

File: examples/demo_bgp.pp [hunk 7]
```diff
@@ -130,46 +130,72 @@
   #---------------------------------------------------------------------------#
```
Comment (chburket) [inline examples/demo_bgp.pp]: @glennmatthews Selectors cleaned this up quite a bit.  I think this is the last change, but I don't want this merged.  Once you are ok with the changes, I'll delete this PR and make a new one to fix the history.

File: examples/demo_bgp.pp [hunk 8]
```diff
@@ -130,46 +130,72 @@
...
     connected_check        => true,
-    capability_negotiation => true,
-    dynamic_capability     => true,
+    capability_negotiation => $capability_negotiation,
+    dynamic_capability     => $dynamic_capability,
     ebgp_multihop          => 2,
-    local_as               => 55.77,
-    log_neighbor_changes   => disable,
-    low_memory_exempt      => false,
+    local_as               => 55.88,
```
Comment (glennmatthews) [inline examples/demo_bgp.pp]: Was this change intentional?

File: examples/demo_bgp.pp (hunk 7 above)
Comment (glennmatthews) [inline examples/demo_bgp.pp]: This looks great to me. Go ahead with the new PR and I'll give it one last once-over today and merge it when done.

File: examples/demo_bgp.pp (hunk 8 above)
Comment (ceburkett) [inline examples/demo_bgp.pp]: Yes.  XR doesn't allow a neighbor local_as to be the same as the BGP AS, so I just changed it for both XR and NX.
